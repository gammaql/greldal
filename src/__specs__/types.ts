import * as t from "io-ts";
import { ExtendsWitness } from "../util-types";
import { MappedDataSource } from "../MappedDataSource";
import { DataSourceMapping } from "../DataSourceMapping";
import { MappedField } from "../MappedField";
import { FieldMapping, FieldMappingArgs } from "../FieldMapping";
import { AssociationMapping } from "../AssociationMapping";

export type _AssociationMappingWitness = ExtendsWitness<
    AssociationMapping<any, any>,
    AssociationMapping<MappedDataSource, MappedDataSource>
>;

export type MappedDataSourceWitness = ExtendsWitness<MappedDataSource<any>, MappedDataSource<DataSourceMapping>>;

export type MappedFieldWitness = ExtendsWitness<MappedField<any, any>, MappedField>;

export type FieldMappingArgsWitness = ExtendsWitness<FieldMappingArgs<any>, FieldMappingArgs<FieldMapping<any, any>>>;

export type FieldMappingWitness = ExtendsWitness<FieldMapping<any, any>, FieldMapping<t.Type<any>, {}>>;
