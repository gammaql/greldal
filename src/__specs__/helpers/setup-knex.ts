import Knex from "knex";
import tmp from "tmp";
import assert from "assert";
import _debug from "debug";

import { Maybe } from "../../util-types";

const debug = _debug("greldal:setup-knex");

export const getConnectionString = () => {
    const connectionString = process.env.DB_CONNECTION_STRING;
    assert(connectionString, "Expected DB_CONNECTION_STRING environment variable to be defined");
    return connectionString;
};

export const setupKnex = () => {
    const isCI = process.env.CI === "true" || process.env.CI === "1";
    let config: Maybe<Knex.Config>;
    const db = process.env.DB;
    if (!db || db === "sqlite3") {
        const sqliteFile = tmp.fileSync();
        config = {
            client: "sqlite3",
            connection: {
                filename: sqliteFile.name,
                multipleStatements: true,
            },
            useNullAsDefault: true,
        };
    } else if (db === "mysql2") {
        if (isCI) {
            config = {
                client: "mysql2",
                connection: {
                    host: "127.0.0.1",
                    user: "travis",
                    database: "greldal_test",
                    password: "",
                },
            };
        } else {
            config = {
                client: "mysql2",
                connection: getConnectionString(),
            };
        }
    } else if (db === "pg") {
        if (isCI) {
            config = {
                client: "pg",
                connection: {
                    user: "postgres",
                    database: "greldal_test",
                },
            };
        } else {
            config = {
                client: "pg",
                connection: getConnectionString(),
            };
        }
    } else {
        throw new Error(`Invalid db selection: ${db}`);
    }
    debug("Connecting to knex using configuration: %O", config);
    assert(config, "Failed to configure database for the test suite. You may need to update above configuration");
    return Knex({
        ...config,
        debug: !!process.env.DEBUG,
        pool: {
            min: 0,
        },
    });
};
