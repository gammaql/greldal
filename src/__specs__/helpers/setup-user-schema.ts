import * as Knex from "knex";
import * as types from "../../types";
import { mapDataSource } from "../../MappedDataSource";
import { GraphQLID, GraphQLString, GraphQLInt } from "graphql";
import { times } from "lodash";
import { mapFields } from "../..";

export const setupUserSchema = async (knex: Knex) => {
    await knex.schema.createTable("users", t => {
        t.increments("id");
        t.string("name");
        t.integer("age");
        t.jsonb("metadata");
    });
};

export const insertFewUsers = async (knex: Knex) => {
    await knex("users").insert([
        {
            id: 1,
            name: "Lorefnon",
            age: 50,
            metadata: JSON.stringify({
                positionsHeld: [
                    {
                        title: "Software Architect",
                        organization: "Foo Bar Inc",
                        duration: 5,
                    },
                    {
                        title: "Software Developer",
                        organization: "Lorem Ipsum Gmbh",
                        duration: 10,
                    },
                ],
                awards: [
                    {
                        title: "Top Achiever",
                        compensation: 1000,
                    },
                ],
            }),
        },
        { id: 2, name: "Gandalf", age: 1000 },
    ]);
};

export const insertManyUsers = async (knex: Knex) => {
    let users = [];
    for (const id of times(500)) {
        users.push({
            id,
            name: `User ${id}`,
            age: 50,
            metadata: JSON.stringify({
                positionsHeld: times(5).map(idx => ({
                    title: `Pos ${id}:${idx}`,
                    organization: `Org ${id}:${idx}`,
                    duration: 3,
                })),
                awards:
                    id > 100
                        ? []
                        : times(3).map(idx => ({
                              title: `Award: ${idx}`,
                              compensation: 100,
                          })),
            }),
        });
        if (users.length >= 20) {
            await knex("users").insert(users);
            users = [];
        }
    }
};

export const mapUsersDataSource = () => {
    // @snippet:start mapDataSource_user_simple
    /// import {mapDataSource, mapFields, types} from "greldal";

    const users = mapDataSource({
        name: "User",
        fields: mapFields({
            id: {
                type: types.number,
                to: GraphQLID,
                isPrimary: true,
            },
            name: {
                type: types.string,
            },
            age: {
                type: types.integer,
            },
        }),
    });
    // @snippet:end
    return users;
};

export const mapUsersDataSourceExplicitly = () => {
    // @snippet:start mapDataSource_user_simple_explicit
    /// import {mapDataSource, mapFields, types} from "greldal";

    const users = mapDataSource({
        name: {
            mapped: "User",
            stored: "users",
        },
        fields: mapFields({
            id: {
                sourceColumn: "id",
                type: types.string,
                to: {
                    input: GraphQLID,
                    output: GraphQLID,
                },
            },
            name: {
                sourceColumn: "name",
                type: types.string,
                to: {
                    input: GraphQLString,
                    output: GraphQLString,
                },
            },
            age: {
                sourceColumn: "age",
                type: types.integer,
                to: {
                    input: GraphQLInt,
                    output: GraphQLInt
                }
            }
        }),
    });
    // @snippet:end
    return users;
};

export const mapUsersDataSourceWithJSONFields = () =>
    mapDataSource({
        name: "User",
        fields: mapFields({
            id: {
                type: types.number,
                to: GraphQLID,
                isPrimary: true,
            },
            name: {
                type: types.string,
            },
            age: {
                type: types.integer,
            },
            metadata: {
                type: types.maybe(
                    types.json(
                        types.partial({
                            positionsHeld: types.array(
                                types.interface({
                                    title: types.string,
                                    organization: types.string,
                                    duration: types.integer,
                                }),
                            ),
                            awards: types.array(
                                types.interface({
                                    compensation: types.number,
                                    title: types.string,
                                }),
                            ),
                        }),
                    ),
                ),
            },
        }),
    });

export const teardownUserSchema = async (knex: Knex) => {
    await knex.schema.dropTable("users");
};
