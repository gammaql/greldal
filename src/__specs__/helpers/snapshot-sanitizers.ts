import { isString, isFunction } from "lodash";
import { isObject } from "util";
import Maybe from "graphql/tsutils/Maybe";

const replaceCode = (err: string) => err.replace(/\[GRelDAL:\S+Error:\d+\]/, "[GRelDAL:ERR_CODE_OMITTED]");

export const removeErrorCodes = (errors: Maybe<any[]>) =>
    (errors || []).map(err => {
        if (isString(err)) return replaceCode(err);
        if (isObject(err) && err.message) return replaceCode(err.message);
        if (isObject(err) && isFunction(err.toString)) return replaceCode(err.toString());
        return replaceCode(`${err}`);
    });
