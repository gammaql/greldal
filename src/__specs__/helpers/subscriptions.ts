import { ExecutionResult } from "graphql";
import { ExecutionResultDataDefault } from "graphql/execution/execute";
import { isFunction } from "lodash";

const isAsyncIterator = (i: any): i is AsyncIterator<any> => i && isFunction(i.next);

export const getSubscriptionResults = (count = 1) => async (
    resultIterator: ExecutionResult<ExecutionResultDataDefault> | AsyncIterator<ExecutionResult<ExecutionResultDataDefault>>
): Promise<ExecutionResult<ExecutionResultDataDefault>[]> => {
    if (!isAsyncIterator(resultIterator)) return [resultIterator];
    const returned: any = [];
    let idx = 0;
    while (true) {
        const { done, value } = await resultIterator.next();
        if (done) return returned;
        returned.push(value);
        idx += 1;
        if (idx === count) return returned;
    }
}