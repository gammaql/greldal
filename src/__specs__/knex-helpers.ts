import * as Knex from "knex";
import { first, values } from "lodash";

export const getCount = async (qb: Knex.QueryBuilder) => Number(first(values(first(await qb.count<any>()))));
