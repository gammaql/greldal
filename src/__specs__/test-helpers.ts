export const reportErrors = async <T>(cb: () => T | Promise<T>): Promise<T> => {
    try {
        return await cb();
    } catch (e) {
        console.error(e);
        throw e;
    }
};
