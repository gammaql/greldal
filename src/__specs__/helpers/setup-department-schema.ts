import * as Knex from "knex";
import { reportErrors } from "../test-helpers";

export const setupDepartmentSchema = async (knex: Knex) => {
    console.log("Setting up department schema: start");
    try {
        await knex.schema.createTable("tags", t => {
            t.increments("id");
            t.string("name");
        });
        await knex.schema.createTable("departments", t => {
            t.increments("id");
            t.string("name");
        });
        await knex.schema.createTable("products", t => {
            t.increments("id");
            t.integer("department_id")
                .unsigned()
                .references("id")
                .inTable("departments")
                .notNullable();
            t.string("name");
        });
        await knex.schema.createTable("product_tag_associators", t => {
            t.increments("id");
            t.integer("product_id")
                .unsigned()
                .references("id")
                .inTable("products")
                .notNullable();
            t.integer("tag_id")
                .unsigned()
                .references("id")
                .inTable("tags")
                .notNullable();
        });
    } catch (e) {
        console.error("Error when setting up departments schema");
        console.error(e);
        throw e;
    }
    console.log("Setting up department schema: end");
};

export const teardownDepartmentSchema = async (knex: Knex) => {
    await reportErrors(async () => {
        await knex.schema.dropTable("product_tag_associators");
        await knex.schema.dropTable("products");
        await knex.schema.dropTable("departments");
        await knex.schema.dropTable("tags");
    });
};
