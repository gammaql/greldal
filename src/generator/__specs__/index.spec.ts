import fs from "fs-extra";
import Knex from "knex";
import path from "path";
import { setupKnex } from "../../__specs__/helpers/setup-knex";
import { useDatabaseConnector } from "../..";
import { generate } from "../index";
import { isEmpty, trim } from "lodash";
import { reportErrors } from "../../__specs__/test-helpers";

jest.setTimeout(60000);

xdescribe("Generator integration", () => {
    let knex: Knex;

    beforeAll(async () => {
        knex = setupKnex();
        useDatabaseConnector(knex);
        const db = process.env.DB || "sqlite3";
        const fixture = await fs.readFile(path.join(__dirname, "__fixtures__", `northwind.${db}.create.sql`));
        for (let stmt of fixture.toString().split(";")) {
            stmt = trim(stmt);
            if (isEmpty(stmt)) continue;
            await knex.schema.raw(stmt);
        }
    });

    afterAll(async () => {
        await reportErrors(async () => {
            for (const tbl of [
                "EmployeeTerritory",
                "CustomerDemographic",
                "CustomerCustomerDemo",
                "OrderDetail",
                "Territory",
                "Region",
                "Product",
                "Order",
                "Supplier",
                "Shipper",
                "Customer",
                "Category",
                "Employee",
            ]) {
                await knex.schema.dropTable(tbl);
            }
            await knex.destroy();
        });
    });

    it("identifies fields and associations", async () => {
        const generated = await generate({ knex });
        expect(generated).toMatchSnapshot();
    });
});
