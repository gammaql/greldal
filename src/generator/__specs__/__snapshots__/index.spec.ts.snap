// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator integration identifies fields and associations 1`] = `
"import { types, mapDataSource, mapFields, mapAssociations } from \\"greldal\\";
const employeeFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.integer,
        isPrimary: true,
    },
    lastName: {
        sourceColumn: \\"LastName\\",
        type: types.string,
    },
    firstName: {
        sourceColumn: \\"FirstName\\",
        type: types.string,
    },
    title: {
        sourceColumn: \\"Title\\",
        type: types.string,
    },
    titleOfCourtesy: {
        sourceColumn: \\"TitleOfCourtesy\\",
        type: types.string,
    },
    birthDate: {
        sourceColumn: \\"BirthDate\\",
        type: types.string,
    },
    hireDate: {
        sourceColumn: \\"HireDate\\",
        type: types.string,
    },
    address: {
        sourceColumn: \\"Address\\",
        type: types.string,
    },
    city: {
        sourceColumn: \\"City\\",
        type: types.string,
    },
    region: {
        sourceColumn: \\"Region\\",
        type: types.string,
    },
    postalCode: {
        sourceColumn: \\"PostalCode\\",
        type: types.string,
    },
    country: {
        sourceColumn: \\"Country\\",
        type: types.string,
    },
    homePhone: {
        sourceColumn: \\"HomePhone\\",
        type: types.string,
    },
    extension: {
        sourceColumn: \\"Extension\\",
        type: types.string,
    },
    photo: {
        sourceColumn: \\"Photo\\",
    },
    note: {
        sourceColumn: \\"Notes\\",
        type: types.string,
    },
    reportsTo: {
        sourceColumn: \\"ReportsTo\\",
        type: types.integer,
    },
    photoPath: {
        sourceColumn: \\"PhotoPath\\",
        type: types.string,
    },
};
const employee = {
    name: {\\"stored\\":\\"Employee\\",\\"mapped\\":\\"Employee\\"},
    fields: mapFields(employeeFields),
};
const categoryFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.integer,
        isPrimary: true,
    },
    categoryName: {
        sourceColumn: \\"CategoryName\\",
        type: types.string,
    },
    description: {
        sourceColumn: \\"Description\\",
        type: types.string,
    },
};
const category = {
    name: {\\"stored\\":\\"Category\\",\\"mapped\\":\\"Category\\"},
    fields: mapFields(categoryFields),
};
const customerFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.string,
        isPrimary: true,
    },
    companyName: {
        sourceColumn: \\"CompanyName\\",
        type: types.string,
    },
    contactName: {
        sourceColumn: \\"ContactName\\",
        type: types.string,
    },
    contactTitle: {
        sourceColumn: \\"ContactTitle\\",
        type: types.string,
    },
    address: {
        sourceColumn: \\"Address\\",
        type: types.string,
    },
    city: {
        sourceColumn: \\"City\\",
        type: types.string,
    },
    region: {
        sourceColumn: \\"Region\\",
        type: types.string,
    },
    postalCode: {
        sourceColumn: \\"PostalCode\\",
        type: types.string,
    },
    country: {
        sourceColumn: \\"Country\\",
        type: types.string,
    },
    phone: {
        sourceColumn: \\"Phone\\",
        type: types.string,
    },
    fax: {
        sourceColumn: \\"Fax\\",
        type: types.string,
    },
};
const customer = {
    name: {\\"stored\\":\\"Customer\\",\\"mapped\\":\\"Customer\\"},
    fields: mapFields(customerFields),
};
const shipperFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.integer,
        isPrimary: true,
    },
    companyName: {
        sourceColumn: \\"CompanyName\\",
        type: types.string,
    },
    phone: {
        sourceColumn: \\"Phone\\",
        type: types.string,
    },
};
const shipper = {
    name: {\\"stored\\":\\"Shipper\\",\\"mapped\\":\\"Shipper\\"},
    fields: mapFields(shipperFields),
};
const supplierFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.integer,
        isPrimary: true,
    },
    companyName: {
        sourceColumn: \\"CompanyName\\",
        type: types.string,
    },
    contactName: {
        sourceColumn: \\"ContactName\\",
        type: types.string,
    },
    contactTitle: {
        sourceColumn: \\"ContactTitle\\",
        type: types.string,
    },
    address: {
        sourceColumn: \\"Address\\",
        type: types.string,
    },
    city: {
        sourceColumn: \\"City\\",
        type: types.string,
    },
    region: {
        sourceColumn: \\"Region\\",
        type: types.string,
    },
    postalCode: {
        sourceColumn: \\"PostalCode\\",
        type: types.string,
    },
    country: {
        sourceColumn: \\"Country\\",
        type: types.string,
    },
    phone: {
        sourceColumn: \\"Phone\\",
        type: types.string,
    },
    fax: {
        sourceColumn: \\"Fax\\",
        type: types.string,
    },
    homePage: {
        sourceColumn: \\"HomePage\\",
        type: types.string,
    },
};
const supplier = {
    name: {\\"stored\\":\\"Supplier\\",\\"mapped\\":\\"Supplier\\"},
    fields: mapFields(supplierFields),
};
const orderFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.integer,
        isPrimary: true,
    },
    customerId: {
        sourceColumn: \\"CustomerId\\",
        type: types.string,
    },
    employeeId: {
        sourceColumn: \\"EmployeeId\\",
        type: types.integer,
    },
    orderDate: {
        sourceColumn: \\"OrderDate\\",
        type: types.string,
    },
    requiredDate: {
        sourceColumn: \\"RequiredDate\\",
        type: types.string,
    },
    shippedDate: {
        sourceColumn: \\"ShippedDate\\",
        type: types.string,
    },
    shipVium: {
        sourceColumn: \\"ShipVia\\",
        type: types.integer,
    },
    freight: {
        sourceColumn: \\"Freight\\",
        type: types.number,
    },
    shipName: {
        sourceColumn: \\"ShipName\\",
        type: types.string,
    },
    shipAddress: {
        sourceColumn: \\"ShipAddress\\",
        type: types.string,
    },
    shipCity: {
        sourceColumn: \\"ShipCity\\",
        type: types.string,
    },
    shipRegion: {
        sourceColumn: \\"ShipRegion\\",
        type: types.string,
    },
    shipPostalCode: {
        sourceColumn: \\"ShipPostalCode\\",
        type: types.string,
    },
    shipCountry: {
        sourceColumn: \\"ShipCountry\\",
        type: types.string,
    },
};
const orderAssociations = {
    employee: {
        target: () => Employee,
        singular: false,
        fetchThrough: [{join: \\"leftOuterJoin\\"}],
        associatorColumns: {
            inSource: \\"EmployeeId\\",
            inRelated: \\"Id\\",
        },
    },
    customer: {
        target: () => Customer,
        singular: false,
        fetchThrough: [{join: \\"leftOuterJoin\\"}],
        associatorColumns: {
            inSource: \\"CustomerId\\",
            inRelated: \\"Id\\",
        },
    },
};
const order = {
    name: {\\"stored\\":\\"Order\\",\\"mapped\\":\\"Order\\"},
    fields: mapFields(orderFields),
    associations: mapAssociations(orderAssociations),
};
const productFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.integer,
        isPrimary: true,
    },
    productName: {
        sourceColumn: \\"ProductName\\",
        type: types.string,
    },
    supplierId: {
        sourceColumn: \\"SupplierId\\",
        type: types.integer,
    },
    categoryId: {
        sourceColumn: \\"CategoryId\\",
        type: types.integer,
    },
    quantityPerUnit: {
        sourceColumn: \\"QuantityPerUnit\\",
        type: types.string,
    },
    unitPrice: {
        sourceColumn: \\"UnitPrice\\",
        type: types.number,
    },
    unitsInStock: {
        sourceColumn: \\"UnitsInStock\\",
        type: types.integer,
    },
    unitsOnOrder: {
        sourceColumn: \\"UnitsOnOrder\\",
        type: types.integer,
    },
    reorderLevel: {
        sourceColumn: \\"ReorderLevel\\",
        type: types.integer,
    },
    discontinued: {
        sourceColumn: \\"Discontinued\\",
        type: types.integer,
    },
};
const product = {
    name: {\\"stored\\":\\"Product\\",\\"mapped\\":\\"Product\\"},
    fields: mapFields(productFields),
};
const orderDetailFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.string,
        isPrimary: true,
    },
    orderId: {
        sourceColumn: \\"OrderId\\",
        type: types.integer,
    },
    productId: {
        sourceColumn: \\"ProductId\\",
        type: types.integer,
    },
    unitPrice: {
        sourceColumn: \\"UnitPrice\\",
        type: types.number,
    },
    quantity: {
        sourceColumn: \\"Quantity\\",
        type: types.integer,
    },
    discount: {
        sourceColumn: \\"Discount\\",
        type: types.number,
    },
};
const orderDetail = {
    name: {\\"stored\\":\\"OrderDetail\\",\\"mapped\\":\\"OrderDetail\\"},
    fields: mapFields(orderDetailFields),
};
const customerCustomerDemoFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.string,
        isPrimary: true,
    },
    customerTypeId: {
        sourceColumn: \\"CustomerTypeId\\",
        type: types.string,
    },
};
const customerCustomerDemo = {
    name: {\\"stored\\":\\"CustomerCustomerDemo\\",\\"mapped\\":\\"CustomerCustomerDemo\\"},
    fields: mapFields(customerCustomerDemoFields),
};
const customerDemographicFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.string,
        isPrimary: true,
    },
    customerDesc: {
        sourceColumn: \\"CustomerDesc\\",
        type: types.string,
    },
};
const customerDemographic = {
    name: {\\"stored\\":\\"CustomerDemographic\\",\\"mapped\\":\\"CustomerDemographic\\"},
    fields: mapFields(customerDemographicFields),
};
const regionFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.integer,
        isPrimary: true,
    },
    regionDescription: {
        sourceColumn: \\"RegionDescription\\",
        type: types.string,
    },
};
const region = {
    name: {\\"stored\\":\\"Region\\",\\"mapped\\":\\"Region\\"},
    fields: mapFields(regionFields),
};
const territoryFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.string,
        isPrimary: true,
    },
    territoryDescription: {
        sourceColumn: \\"TerritoryDescription\\",
        type: types.string,
    },
    regionId: {
        sourceColumn: \\"RegionId\\",
        type: types.integer,
    },
};
const territory = {
    name: {\\"stored\\":\\"Territory\\",\\"mapped\\":\\"Territory\\"},
    fields: mapFields(territoryFields),
};
const employeeTerritoryFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.string,
        isPrimary: true,
    },
    employeeId: {
        sourceColumn: \\"EmployeeId\\",
        type: types.integer,
    },
    territoryId: {
        sourceColumn: \\"TerritoryId\\",
        type: types.string,
    },
};
const employeeTerritory = {
    name: {\\"stored\\":\\"EmployeeTerritory\\",\\"mapped\\":\\"EmployeeTerritory\\"},
    fields: mapFields(employeeTerritoryFields),
};
const productDetailsVFields = {
    id: {
        sourceColumn: \\"Id\\",
        type: types.integer,
    },
    productName: {
        sourceColumn: \\"ProductName\\",
        type: types.string,
    },
    supplierId: {
        sourceColumn: \\"SupplierId\\",
        type: types.integer,
    },
    categoryId: {
        sourceColumn: \\"CategoryId\\",
        type: types.integer,
    },
    quantityPerUnit: {
        sourceColumn: \\"QuantityPerUnit\\",
        type: types.string,
    },
    unitPrice: {
        sourceColumn: \\"UnitPrice\\",
        type: types.number,
    },
    unitsInStock: {
        sourceColumn: \\"UnitsInStock\\",
        type: types.integer,
    },
    unitsOnOrder: {
        sourceColumn: \\"UnitsOnOrder\\",
        type: types.integer,
    },
    reorderLevel: {
        sourceColumn: \\"ReorderLevel\\",
        type: types.integer,
    },
    discontinued: {
        sourceColumn: \\"Discontinued\\",
        type: types.integer,
    },
    categoryName: {
        sourceColumn: \\"CategoryName\\",
        type: types.string,
    },
    categoryDescription: {
        sourceColumn: \\"CategoryDescription\\",
        type: types.string,
    },
    supplierName: {
        sourceColumn: \\"SupplierName\\",
        type: types.string,
    },
    supplierRegion: {
        sourceColumn: \\"SupplierRegion\\",
        type: types.string,
    },
};
const productDetailsV = {
    name: {\\"stored\\":\\"ProductDetails_V\\",\\"mapped\\":\\"ProductDetailsV\\"},
    fields: mapFields(productDetailsVFields),
};
"
`;
